derive/src/parse.rs

```diff
@@ -458,103 +482,150 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
         name: enum_name,
         variants,
         attributes: vec![],
-        generics: generic_types
+        generics: generic_types.into_iter().collect(),
     }
 }
 
-fn get_bounds(source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> HashMap<String, String> {
-    let mut ret = HashMap::new();
+fn get_bounds(
+    source: &mut Peekable<impl Iterator<Item = TokenTree>>,
+) -> Vec<(String, Vec<String>)> {
+    let mut ret = Vec::new();
 
+    // Angle bracket generics + bounds
     match source.peek() {
         Some(content) if content.to_string() == "<" => {
             source.next();
-            let mut typename = "".to_string(); 
-            let mut generic_bound = String::new();
+            let mut typename: Option<String> = None;
+            let mut boundname: Option<String> = None;
+            let mut generic_bounds = Vec::new();
             let mut in_type = true;
             while let Some(tok) = source.next() {
                 match tok.to_string().as_str() {
                     ">" => {
-                        ret.insert(typename.clone(), generic_bound.clone());
-                        typename.clear();
-                        generic_bound.clear();
-                        break
-                    },
-                    ":" => {
+                        if let Some(boundname) = boundname.take() {
+                            generic_bounds.push(boundname)
```

Comment:
I am probably missing something, but why can't we push to the `ret` right away? It looks like we always have the type first and bounds second, so I don't see a reason for this second buffer?

Reply:
Basically just for clarity. It's easier to think of each type with its own list of individual bounds. Nanoserde didn't support generics at all before, so was mostly just trying to get everything working without worrying about performance too much. It's still very fast, I'm using this branch in my own code and it hasn't really changed compile times at all.